### PR TITLE
feat(mongoose): Document Connection.then/catch

### PR DESCRIPTION
**Pull request by ethanresnick**
If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<https://github.com/Automattic/mongoose/blob/master/lib/index.js#L155>> <<https://github.com/Automattic/mongoose/blob/master/lib/connection.js#L489>>
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

This picks up where https://github.com/DefinitelyTyped/DefinitelyTyped/pull/25885 left off

### CODE REVIEW

File: types/mongoose/index.d.ts [hunk 1]
```diff
@@ -403,6 +403,9 @@ declare module "mongoose" {
 
     /** Expose the possible connection states. */
     static STATES: any;
+
+    then: Promise<Connection>["then"];
+    catch: Promise<Connection>["catch"];
```
[the remark this answers is not in the document]
Comment (ethanresnick) [reply types/mongoose/index.d.ts]: @DanielRosenwasser Yeah, but Connection actually doesn't have `Promise.prototype.finally` on it. (The Mongoose code manually attaches only `then` and `catch`.)

**Comment by flisky** [comment]
It breaks my code:

```
const getDB = async (tenant: string): Promise<Connection> => {
  await ...
  return sysDB.useDb(tenant);
};
```
gives me `error TS1062: Type is referenced directly or indirectly in the fulfillment callback of its own 'then' method.`

Any suggestions? Thanks!

**Comment by ethanresnick** [comment]
@flisky Oy, I see the error here. Basically, the Mongoose docs for `createConnection` say that a Connection object is thenable, but really it isn't in the general case; instead, `then`/`catch` methods are only added to the connection objects returned from specific methods (namely, `Connection.prototype.openUri`, which is used by `createConnection`). If `Connection`'s were thenable in the general case, TS would be right to give you the error that you're seeing, because (I think) the await would trigger an infinite loop. I'll open a PR to fix this.

**Comment by ethanresnick** [comment]
@flisky See https://github.com/DefinitelyTyped/DefinitelyTyped/pull/26509, which should fix this.

**Comment by flisky** [comment]
Thank you for the fix :100:

**Pull request merged**